src/contrib/persistence/Akka.Persistence.Sql.Common/Journal/BatchingSqlJournal.cs

```diff
@@ -1388,29 +1389,111 @@ protected void AddParameter(TCommand command, string paramName, DbType dbType, o
         /// <param name="param">Parameter to customize</param>
         protected virtual void PreAddParameterToCommand(TCommand command, DbParameter param) { }
         
+        /// <summary>
+        /// Select the buffer that has the smallest id on its first item, retrieve a maximum Setup.MaxBatchSize
+        /// items from it, and return it as a chunk that needs to be batched
+        /// </summary>
         private RequestChunk DequeueChunk(int chunkId)
         {
-            var operationsCount = Math.Min(Buffer.Count, Setup.MaxBatchSize);
-            var array = new IJournalRequest[operationsCount];
-            for (int i = 0; i < operationsCount; i++)
+            var currentBuffer = _buffers
+                .Where(q => q.Count > 0)
+                .OrderBy(q => q.Peek().id).First();
+
+            var operations = new List<IJournalRequest>();
+            if (ReferenceEquals(currentBuffer, WriteBuffer))
+            {
+                // Stop dequeuing when we encounter another type of write operation request
+                // We don't batch delete and writes in the same batch, reason being a database
+                // can be deadlocked if write and delete happens in the same transaction
+                var writeType = currentBuffer.Peek().request.GetType();
+                while(currentBuffer.Count > 0 && currentBuffer.Peek().request.GetType() == writeType)
+                {
+                    operations.Add(currentBuffer.Dequeue().request);
+                    if (operations.Count == Setup.MaxBatchSize)
+                        break;
+                }
+            }
+            else
             {
-                var req = Buffer.Dequeue();
-                array[i] = req;
+                while(currentBuffer.Count > 0)
+                {
+                    operations.Add(currentBuffer.Dequeue().request);
+                    if (operations.Count == Setup.MaxBatchSize)
+                        break;
+                }
             }
-
-            return new RequestChunk(chunkId, array);
+            
+            return new RequestChunk(chunkId, operations.ToArray());
```

Comment:
Why not just pass the `List<>` instead?

Reply:
RequestChunk is a struct, I'm not sure if there is a specific reason why it is using an array and not a list, I assumed that it is for optimization reason, so I didn't change it.